Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -253,3 +253,176 @@ theorem Language.IsContextFree.reverse (L : Language T) :
   fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
 
 end closure_reversal
+
+section embed_project
+
+/-! This section contains only auxiliary constructions that will shorten upcoming proofs of
+closure properties. When combining several grammars together, we usually want to take a sum type of
+their nonterminal types and embed respective nonterminals to this sum type. We subsequently show
+that the resulting grammar preserves derivations of those strings that may contain any terminals but
+only the proper nonterminals. The embedding operation must be injective. The projection operation
+must be injective on those symbols where it is defined. -/
+
+/-- Mapping `Symbol` when it is a nonterminal. -/
+def Symbol.map {N₀ N : Type*} (f : N₀ → N) : Symbol T N₀ → Symbol T N
+  | Symbol.terminal t => Symbol.terminal t
+  | Symbol.nonterminal n => Symbol.nonterminal (f n)
+
+/-- Mapping `Symbol` when it is a nonterminal; may return `none`. -/
+def Symbol.filterMap {N₀ N : Type*} (f : N → Option N₀) : Symbol T N → Option (Symbol T N₀)
+  | Symbol.terminal t => some (Symbol.terminal t)
+  | Symbol.nonterminal n => Option.map Symbol.nonterminal (f n)
+
+/-- Map the type of nonterminal symbols of a `ContextFreeRule` . -/
+def ContextFreeRule.map {N₀ N : Type*} (r : ContextFreeRule T N₀) (f : N₀ → N) :
+    ContextFreeRule T N :=
+  ⟨f r.input, r.output.map (Symbol.map f)⟩
+
+/-- A pair of `ContextFreeGrammar`s with maps between their nonterminal types that work,
+roughly speaking, in a good way. -/
+structure EmbeddedContextFreeGrammar (T : Type uT) where
+  /-- The smaller grammar. -/
+  g₀ : ContextFreeGrammar.{uN} T
+  /-- The bigger grammar. -/
+  g : ContextFreeGrammar.{uN} T
+  /-- Mapping nonterminals from the smaller type to the bigger type. -/
+  embedNT : g₀.NT → g.NT
+  /-- Mapping nonterminals from the bigger type to the smaller type. -/
+  projectNT : g.NT → Option g₀.NT
+  /-- The former map is injective. -/
```

Comment:
You shouldn't rely on the order of the fields in docstrings
```suggestion
  /-- The function on nonterminal symbols is injective. -/
```

Reply:
Both are functions non nonterminal symbols.

Reply:
```suggestion
  /-- The embedding map is injective. -/
```